File: services/engine/app/reconcile.py
"""Running-balance reconciliation + confidence scoring.

This is the product's accuracy story and trust badge: for every transaction we
assert `previous_balance + money_in - money_out == stated_balance`. Rows that
break the chain are flagged (not silently "fixed"), and the overall confidence
is the share of checkable rows that reconcile. All arithmetic is in integer
pence.
"""
from __future__ import annotations

from typing import Optional

from .config import RECONCILE_TOLERANCE_PENCE
from .normalize import pence_to_pounds
from .parser import RawRow
from .schema import ReconcileSummary, Transaction


def _movement(row: RawRow) -> int:
    return (row.money_in or 0) - (row.money_out or 0)


def reconcile(
    raw_rows: list[RawRow], tolerance_pence: int = RECONCILE_TOLERANCE_PENCE
) -> tuple[list[Transaction], ReconcileSummary, float]:
    txns = [r for r in raw_rows if r.kind == "txn"]
    opening_row = next((r for r in raw_rows if r.kind == "opening"), None)
    closing_row = next((r for r in raw_rows if r.kind == "closing"), None)

    # Establish the opening balance to anchor the chain.
    opening: Optional[int] = None
    opening_backcomputed = False
    if opening_row is not None:
        opening = opening_row.balance
    elif txns and txns[0].balance is not None:
        # Back-compute the opening from the first row's stated balance. This
        # makes the first row reconcile by construction, so it is NOT an
        # independent check — we exclude it from the confidence denominator.
        opening = txns[0].balance - _movement(txns[0])
        opening_backcomputed = True

    out_rows: list[Transaction] = []
    prev: Optional[int] = opening
    checkable = 0
    reconciled_count = 0

    for idx, row in enumerate(txns):
        movement = _movement(row)
        reconciled = True
        flag: Optional[str] = None
        balance = row.balance

        if balance is None:
            # Statement should print a balance; infer it but flag for review.
            if prev is not None:
                balance = prev + movement
                reconciled = False
                flag = "Balance not printed — inferred from running total"
            prev = balance
        elif prev is None:
            # No anchor yet: first row sets the baseline, nothing to verify.
            prev = balance
        else:
            expected = prev + movement
            # The back-computed opening guarantees row 0 ties out; don't let that
            # synthetic pass inflate confidence.
            counts = not (opening_backcomputed and idx == 0)
            if counts:
                checkable += 1
            if abs(expected - balance) <= tolerance_pence:
                if counts:
                    reconciled_count += 1
            else:
                reconciled = False
                diff = pence_to_pounds(balance - expected)
                exp = pence_to_pounds(expected)
                flag = (
                    f"Running balance off by £{diff:+.2f} "
                    f"(expected £{exp:.2f}, statement shows £{pence_to_pounds(balance):.2f})"
                )
            prev = balance  # continue from the stated balance, not our guess

        out_rows.append(
            Transaction(
                date=row.date_iso or row.date_raw,
                description=row.description,
                money_in=pence_to_pounds(row.money_in),
                money_out=pence_to_pounds(row.money_out),
                balance=pence_to_pounds(balance),
                reconciled=reconciled,
                flag_reason=flag,
            )
        )

    # Closing reconciliation.
    stated_closing = (
        closing_row.balance
        if closing_row is not None
        else (txns[-1].balance if txns and txns[-1].balance is not None else None)
    )
    computed_closing = None
    if opening is not None:
        computed_closing = opening + sum(_movement(r) for r in txns)
    balanced = (
        stated_closing is not None
        and computed_closing is not None
        and abs(stated_closing - computed_closing) <= tolerance_pence
    )

    flagged_rows = sum(1 for t in out_rows if not t.reconciled)
    summary = ReconcileSummary(
        total_rows=len(out_rows),
        movement_rows=len(txns),
        reconciled_rows=len(out_rows) - flagged_rows,
        flagged_rows=flagged_rows,
        opening_balance=pence_to_pounds(opening),
        closing_balance=pence_to_pounds(stated_closing),
        computed_closing=pence_to_pounds(computed_closing),
        balanced=balanced,
    )

    confidence = _confidence(checkable, reconciled_count, len(out_rows), balanced)
    return out_rows, summary, confidence


def _confidence(
    checkable: int, reconciled_count: int, total_rows: int, balanced: bool
) -> float:
    if total_rows == 0:
        return 0.0
    if checkable == 0:
        # We extracted rows but had nothing to verify against (no anchor).
        return 0.3
    ratio = reconciled_count / checkable
    # Small bonus when the end-to-end opening→closing also ties out.
    if balanced and ratio >= 0.999:
        return 1.0
    return round(min(ratio, 0.99), 4)

File: apps/tabular/lib/reconcile.ts
// Client-side running-balance reconciliation — mirrors
// services/engine/app/reconcile.py so the trust signals (per-row flags,
// summary, confidence) stay live and honest as the user edits cells.
//
// The product's promise is "every row is verified against the statement's own
// running balance." That verification must therefore re-run on every edit, not
// be frozen at extraction time. All arithmetic is in integer pence to stay
// exact, exactly like the engine.

import type { ReconcileSummary, Transaction } from "@/lib/types";

// Matches ENGINE_RECONCILE_TOLERANCE_PENCE (1p) in services/engine/app/config.py.
const TOLERANCE_PENCE = 1;

function toPence(pounds: number | null): number | null {
  return pounds == null ? null : Math.round(pounds * 100);
}

function toPounds(pence: number | null): number | null {
  return pence == null ? null : Math.round(pence) / 100;
}

function movementPence(t: Transaction): number {
  return (toPence(t.money_in) ?? 0) - (toPence(t.money_out) ?? 0);
}

export interface ReconcileResult {
  rows: Transaction[];
  summary: ReconcileSummary;
  confidence: number;
}

/**
 * Re-derive reconciliation for the (possibly edited) rows.
 *
 * `anchorOpening` is the opening balance established at extraction time (from
 * the statement's "balance brought forward" line). When absent we back-compute
 * it from the first row, mirroring the engine.
 */
export function reconcile(
  rows: Transaction[],
  anchorOpening: number | null,
): ReconcileResult {
  let opening = toPence(anchorOpening);
  const first = rows[0];
  let openingBackcomputed = false;
  if (opening == null && first != null && first.balance != null) {
    opening = (toPence(first.balance) ?? 0) - movementPence(first);
    openingBackcomputed = true;
  }

  const out: Transaction[] = [];
  let prev: number | null = opening;
  let checkable = 0;
  let reconciledCount = 0;

  for (let idx = 0; idx < rows.length; idx++) {
    const row = rows[idx]!;
    const movement = movementPence(row);
    let reconciled = true;
    let flag: string | null = null;
    let balance = toPence(row.balance);

    if (balance == null) {
      if (prev != null) {
        balance = prev + movement;
        reconciled = false;
        flag = "Balance not printed — inferred from running total";
      }
      prev = balance;
    } else if (prev == null) {
      // No anchor yet: first row sets the baseline, nothing to verify.
      prev = balance;
    } else {
      const expected = prev + movement;
      // The back-computed opening guarantees row 0 ties out; don't let that
      // synthetic pass inflate confidence.
      const counts = !(openingBackcomputed && idx === 0);
      if (counts) checkable += 1;
      if (Math.abs(expected - balance) <= TOLERANCE_PENCE) {
        if (counts) reconciledCount += 1;
      } else {
        reconciled = false;
        const diff = toPounds(balance - expected)!;
        const exp = toPounds(expected)!;
        const sign = diff >= 0 ? "+" : "";
        flag =
          `Running balance off by £${sign}${diff.toFixed(2)} ` +
          `(expected £${exp.toFixed(2)}, statement shows £${toPounds(balance)!.toFixed(2)})`;
      }
      prev = balance; // continue from the stated balance, not our guess
    }

    out.push({ ...row, balance: toPounds(balance), reconciled, flag_reason: flag });
  }

  const last = rows[rows.length - 1];
  const statedClosing = last ? toPence(last.balance) : null;
  const computedClosing =
    opening != null
      ? opening + rows.reduce((acc, r) => acc + movementPence(r), 0)
      : null;
  const balanced =
    statedClosing != null &&
    computedClosing != null &&
    Math.abs(statedClosing - computedClosing) <= TOLERANCE_PENCE;

  const flaggedRows = out.filter((t) => !t.reconciled).length;
  const summary: ReconcileSummary = {
    total_rows: out.length,
    movement_rows: out.length,
    reconciled_rows: out.length - flaggedRows,
    flagged_rows: flaggedRows,
    opening_balance: toPounds(opening),
    closing_balance: toPounds(statedClosing),
    computed_closing: toPounds(computedClosing),
    balanced,
  };

  return { rows: out, summary, confidence: confidence(checkable, reconciledCount, out.length, balanced) };
}

function confidence(
  checkable: number,
  reconciledCount: number,
  totalRows: number,
  balanced: boolean,
): number {
  if (totalRows === 0) return 0;
  if (checkable === 0) return 0.3;
  const ratio = reconciledCount / checkable;
  if (balanced && ratio >= 0.999) return 1;
  return Math.round(Math.min(ratio, 0.99) * 10000) / 10000;
}
